zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java

```diff
@@ -0,0 +1,297 @@
+        final long timeBucket = TableHelper.getTimeBucket(table);
+        final String tagTable = TableHelper.getTable(ZipkinSpanRecord.ADDITIONAL_QUERY_TABLE, timeBucket);
+        for (Map.Entry<String, String> entry : request.annotationQuery().entrySet()) {
+          completionTraceIds.add(client.executeAsyncQuery("select " + ZipkinSpanRecord.TRACE_ID + " from " + tagTable +
+                  " where " + ZipkinSpanRecord.QUERY + " = ?" +
+                  " and " + ZipkinSpanRecord.TIME_BUCKET + " >= ?" +
+                  " and " + ZipkinSpanRecord.TIME_BUCKET + " <= ? ALLOW FILTERING",
+              row -> row.getString(ZipkinSpanRecord.TRACE_ID),
+              entry.getValue().isEmpty() ? entry.getKey() : entry.getKey() + "=" + entry.getValue(),
+              duration.getStartTimeBucket(), duration.getEndTimeBucket()));
+        }
+      }
+      if (request.minDuration() != null) {
+        completionTraceIds.add(client.executeAsyncQuery("select " + ZipkinSpanRecord.TRACE_ID + " from " + table +
+                " where " + ZipkinSpanRecord.DURATION + " >= ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " >= ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " <= ? ALLOW FILTERING",
+            row -> row.getString(ZipkinSpanRecord.TRACE_ID),
+            request.minDuration(), duration.getStartTimeBucket(), duration.getEndTimeBucket()
+        ));
+      }
+      if (request.maxDuration() != null) {
+        completionTraceIds.add(client.executeAsyncQuery("select " + ZipkinSpanRecord.TRACE_ID + " from " + table +
+                " where " + ZipkinSpanRecord.DURATION + " <= ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " >= ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " <= ? ALLOW FILTERING",
+            row -> row.getString(ZipkinSpanRecord.TRACE_ID),
+            request.maxDuration(), duration.getStartTimeBucket(), duration.getEndTimeBucket()
+        ));
+      }
+      if (StringUtil.isNotEmpty(request.serviceName())) {
+        completionTraceIds.add(client.executeAsyncQuery("select " + ZipkinSpanRecord.TRACE_ID + " from " + table +
+                " where " + ZipkinSpanRecord.LOCAL_ENDPOINT_SERVICE_NAME + " = ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " >= ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " <= ? ALLOW FILTERING",
+            row -> row.getString(ZipkinSpanRecord.TRACE_ID),
+            request.serviceName(), duration.getStartTimeBucket(), duration.getEndTimeBucket()
+        ));
+      }
+      if (StringUtil.isNotEmpty(request.remoteServiceName())) {
+        completionTraceIds.add(client.executeAsyncQuery("select " + ZipkinSpanRecord.TRACE_ID + " from " + table +
+                " where " + ZipkinSpanRecord.REMOTE_ENDPOINT_SERVICE_NAME + " = ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " >= ?" +
+                " and " + ZipkinSpanRecord.TIME_BUCKET + " <= ? ALLOW FILTERING",
```

Comment:
is the primary key specified in the where clause here? 
allow filtering without the primary key specified is a full-table scan and will not work.